### ISSUE TITLE
Flow seems to stop typing exact types in some conditions

### ISSUE THREAD

**Issue by buoyad** [facebook/flow#5429]
TL;DR, See minimal example [here](https://flow.org/try/#0C4TwDgpgBAggxsAlgewHZQLxQN4CgCQYAhiADbJEAmAXDgD65RNT6VHBG0Ai7RAKuAiNmdAL65xuUJCg8OAmVmwB6AFRQACiXJUobDlADuEAOQAnaIlTAIFgM43KUK1FXLJcNA6h2iAcyJMKAAKIgQUVFp4JDQASkwAPhwCUghgKGIyCicsMJjUADpMnUphJkQAMxDi7IKAV1QAa1RkQ1QNM2RIM1B4vGYWT1RvVAhDaIig-oHmfFHDOU4M7VrkYAALWw6u21AymfEB8VEgA)

[This](https://flow.org/try/#0C4TwDgpgBAggxsAlgewHZQLxQN4CgCQYAhiADbJEAmAXDrlA1PpUcEbQCKtEAq4E9RgF9cI3KEhQubPpKzYA9ACooABRLkqUFmygB3CAHIATtESpgEUwGdLlKOahKFYuGltRrRAOZFMUAAoiBBRUWngkNABKTAA+OnxSCGAoYjIKeyxgyNQAOjTNSkEGRAAzQIKM3IBXVABrVGQ9VFVjZEhjUBi8RiY3VA9UCD0I0P8e3sZ8Ib1pdlSNKuRgAAsrVvarUGLJkV6RISA) first example produces the expected flow behavior:

```js
type Action = {
	payload: {
    	    data: DataType
        }
}

type DataType = {/* Payload data we're interested in */}

const saga = (action: Action) => {
    let payload = action.payload
    if (payload.unknownProperty) {
    	const newAction = {
        	newData: payload.otherProperty // Error: property `otherProperty`. Property not found in object type
        }
    }
}
```

`if (payload.unknownProperty)` only guarantees that `unknownProperty` exists. When we try to access a totally unrelated `otherProperty`, flow complains because there's no refinement that says it exists.

However, if `Action.payload` is an exact type...

```js
type Action = {
	payload: {|
    	    data: DataType
        |}
}

type DataType = {/* Payload data we're interested in */}

const saga = (action: Action) => {
    let payload = action.payload
    if (payload.unknownProperty) {
    	const newAction = {
        	newData: payload.otherProperty // No error / no typing?
        }
    }
}
```
Flow says `No errors!` and the type indicator in the [playground](https://flow.org/try/#0C4TwDgpgBAggxsAlgewHZQLxQN4CgCQYAhiADbJEAmAXDgD65RNT6VHBG0Ai7RAKuAiNmdAL65xuUJCg8OAmVmwB6AFRQACiXJUobDlADuEAOQAnaIlTAIFgM43KUK1FXLJcNA6h2iAcyJMKAAKIgQUVFp4JDQASkwAPhwCUghgKGIyCicsMJjUADpMnUphJkQAMxDi7IKAV1QAa1RkQ1QNM2RIM1B4vGYWT1RvVAhDaIig-oHmfFHDOU4M7VrkYAALWw6u21AymfEB8VEgA) doesn't say anything about the line with `payload.otherProperty`. I'm not sure if this is expected behavior, but I noticed it debugging another issue that it looks like flow could have warned me about.

Thanks!

**Comment by rvion** [comment]
might be related to https://github.com/facebook/flow/issues/5406

**Comment by samwgoldman** [comment]
This is actually the expected behavior. Because `Flow` knows that `payload.unknownProperty` is always undefined, the entire `if`-statement is dead code, so there can't be any runtime errors.

**Comment by chrisnojima** [comment]
Why doesn't flow complain about accessing that undefined variable?

**Comment by samwgoldman** [comment]
Flow allows you to test for properties like this in conditionals. It's a common pattern for property testing and also refinements. Consider a union of exact objects; the property test can refine the union to one branch or the other.

**Comment by buoyad** [comment]
Based on your first comment @samwgoldman, it seems that flow knows that `payload.unknownProperty` doesn't exist and marks the conditional as dead code (since `Action` is an exact type). Given that, shouldn't flow give an error at `if (payload.unknownProperty)`?

**Comment by apsavin** [comment]
>Flow allows you to test for properties like this in conditionals.

Previous comment

**Comment by buoyad** [comment]
Thanks, I read that comment. It doesn't really address this part of my comment:

> it seems that flow knows that payload.unknownProperty doesn't exist

It seems to me that either flow should complain at the conditional in this case (or the line after with 'dead code') or complain at the access of `action.otherProperty` like it does with the non-exact type. I think not complaining at all can allow errors to pass through (especially because this only happens with the exact type), like the issue I referenced at the beginning of this thread.

**Comment by akalin-keybase** [comment]
This behavior with exact types, although it makes sense after having it explained, is highly surprising. I think most people expect that making a type exact would increase the number of warnings given, instead of decreasing.
